source/dotnet/Library/AdaptiveCards.Templating/AdaptiveCardsTemplateVisitor.cs

```diff
@@ -218,11 +223,15 @@ public override AdaptiveCardsTemplateResult VisitTemplateData([NotNull] Adaptive
                     }
                     catch (ArgumentNullException e)
                     {
-                        throw new Exception($"'{templateLiteral.Symbol.Text}' at line, '{templateLiteral.Symbol.Line}' is invalid because '{e.Message}'");
+                        throw new AdaptiveTemplateException($"'{templateLiteral.Symbol.Text}' at line, '{templateLiteral.Symbol.Line}' is invalid because '{e.Message}'");
                     }
-                    catch (Exception e)
+                    catch (JsonException e)
+                    {
+                        throw new AdaptiveTemplateException($"'{templateLiteral.Symbol.Text}' at line, '{templateLiteral.Symbol.Line}' is malformed for '$data : ' pair", e);
```

Comment:
How are you distinguishing "invalid" vs ""malformed" as far as the customer is concerned? Is this an error in the template or the data? Is JSONexception for the for the data portion? If so we should probably correlate it explicitly in the error message we are forming out for the exception..

Reply:
if parsing can be initiated, and failed, then it's malformed. if parsing could even be started such as null selection path, then, it's invalid.

---
In reply to: [426855989](https://github.com/microsoft/AdaptiveCards/pull/4026#discussion_r426855989) [](ancestors = 426855989)

Reply:
Ok thanks. So it almost feels like we can simply alter the string then to be more actionable for the user by including this bit of context int the error message for the ArgumentNullException handling and maybe bubble up the ArgumentNullException as-is to distinguish between teh 2 scenarios explicitly - ([https://docs.microsoft.com/en-us/dotnet/standard/exceptions/](https://docs.microsoft.com/en-us/dotnet/standard/exceptions/))

How about this? 

try
{
 ...
}
catch(ArgumentNullException)
{
   var errorMessage = $"Please enter a non-null value for '{templateLiteral.Symbol.Text}' at line, '{templateLiteral.Symbol.Line}''"; // Do we need to relay the innerException.Message? What more context is it providing to the user? 
   throw new ArgumentNullException(errorMessage);
}
catch(JsonException innerEx)
{
  throw new AdaptiveTemplateException($"'{templateLiteral.Symbol.Text}' at line, '{templateLiteral.Symbol.Line}' is malformed for '$data : ' pair", innerException);
}


---
In reply to: [426858661](https://github.com/microsoft/AdaptiveCards/pull/4026#discussion_r426858661) [](ancestors = 426858661,426855989)

Reply:
1. Nit: i used a string variable in my suggestion since that's my preference -- if you want to do it, then use the same format for the handler below too for consistency; 
2. We should extend your unit tests to cover both these cases so we can use that later for sample output/clarifying the behavior in our documentation also.. Wll we need to regenerate the API docs here? 


---
In reply to: [426946169](https://github.com/microsoft/AdaptiveCards/pull/4026#discussion_r426946169) [](ancestors = 426946169,426858661,426855989)